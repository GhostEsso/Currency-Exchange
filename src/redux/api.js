import { createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const FetchData = createAsyncThunk('fetching/app', async () => {
  try {
    const res = await axios.get('https://api.currencyfreaks.com/v2.0/supported-currencies');
    const data = await res.data;
    return data.supportedCurrenciesMap;
  } catch (error) {
    return error;
  }
});

const GetData = createAsyncThunk('getdata/app', async ({ amount, from, to }) => {
  const res = await axios.get(`https://api.exchangerate.host/convert?from=${from}&to=${to}&amount=${amount}`);
  const data = await res.data;
  return data;
});

const LatestData = createAsyncThunk('latest/currency', async () => {
  const res = await axios.get('https://api.exchangerate.host/latest?base=USD');
  const data = await res.data;
  return data;
});
export { GetData, FetchData, LatestData };
