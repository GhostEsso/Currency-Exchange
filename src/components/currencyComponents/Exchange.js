import { Link } from 'react-router-dom';
import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { FetchData } from '../../redux/api';
import styles from '../Styles.module.css';

const Card = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(FetchData());
  }, [dispatch]);

  const { exchange, filters } = useSelector((state) => state.exchange);
  const Limit = Object.values(exchange).slice(0, 18);

  if (filters && filters.length > 0) {
    return (
      <Link
        to={`/support/${filters[0].currencyCode}`}
        key={filters[0].currencyCode}
        className={styles.styleLink}
      >
        <div className="CardContainer">
          <div className="cards">
            <img src={filters[0].icon} alt="" />
            <div className="currency">
              <h1>Currency Names:</h1>
              <p>{filters[0].currencyName}</p>
            </div>
            <div className="date">
              <p>
                {filters[0].availableFrom}
                {' '}
                -
                {filters[0].availableUntil}
              </p>
            </div>
          </div>
        </div>
      </Link>
    );
  }
  return (
    <div className="Containers">
      {Limit.map((currency) => (
        <Link
          to={`/support/${currency.currencyCode}`}
          key={currency.currencyCode}
          className={styles.styleLink}
        >
          <div className="card">
            <img src={currency.icon} alt="" />
            <div className="currency">
              <h1>Currency Names:</h1>
              <p>{currency.currencyCode}</p>
            </div>
            <div className="date">
              <p>
                {currency.availableFrom}
                {' '}
                -
                {currency.availableUntil}
              </p>
            </div>
          </div>
        </Link>
      ))}
    </div>
  );
};

export default Card;
