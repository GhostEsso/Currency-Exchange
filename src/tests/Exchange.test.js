import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import axios from 'axios';
import { Provider } from 'react-redux';
import Exchange from './Exchange';
import { FetchData } from './api';

// Mock axios module
jest.mock('axios', () => ({
  get: jest.fn(() => Promise.resolve({ data: {} })),
}));

// Create a mock Redux store with thunk middleware
const mockStore = configureStore([thunk])({
  exchange: {
    exchange: {},
    supported: {
      symbols: {
        USD: {},
        AED: {},
      },
    },
  },
});

describe('Exchange Display', () => {
  it('should render Exchange component correctly', () => {
    // Arrange
    const component = (
      <Provider store={mockStore}>
        <BrowserRouter>
          <Exchange />
        </BrowserRouter>
      </Provider>
    );

    // Act
    const tree = renderer.create(component).toJSON();

    // Assert
    expect(tree).toMatchSnapshot();
  });

  it('should call FetchData with the correct URL', async () => {
    // Arrange
    const expectedURL = 'https://api.currencyfreaks.com/v2.0/supported-currencies';

    // Act
    await mockStore.dispatch(FetchData());

    // Assert
    expect(axios.get).toHaveBeenCalledWith(expectedURL);
  });
});
