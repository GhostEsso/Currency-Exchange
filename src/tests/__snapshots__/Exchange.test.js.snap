// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange Display renders correctly 1`] = `
<div
  className="Containers"
/>
`;
